### DISCUSSION TITLE
Lessons learned from performance optimization - an unlikely culprit

### DISCUSSION THREAD

**Discussion by suzdalnitski** [commanded/commanded#646] [Show and tell]
I've been profiling the app I'm working on, and discovered that dispatching commands to Commanded was taking way too long - about 80ms per dispatch... 

I was hoping that this would get better with parallelization, but I saw **zero** improvements. Running 1000 dispatches in parallel (100 parallel processes) still resulted in the same 75-80ms per dispatch (where ideally, it should have gone down x100-fold).

My initial suspicion was the EventStore - after all this is the only point where we interact with the database (other than read model projections, but they were super quick already). Database often is the bottleneck in many apps. I've ran a [benchmark](https://github.com/commanded/eventstore/issues/256) on appending events to the EventStore, and discovered that at 1 million events, appending only takes about 75 uSecs per append (it does get slower though with time). This means that EventStore can't be the culprit.

My next step was to instrument the Commanded library itself with `:timer.tc`, and I found that the following function calls were the slowest:

**No parallelism:**
```
[88,704  us] chain [before_dispatch] Elixir.MyApp.Commanded.Support.Middleware.Validate
[10,204  us] chain [after_dispatch] Elixir.Commanded.Middleware.ConsistencyGuarantee
[3,318   us] Aggregate.persist_events.append_to_stream 
```

**10x parallelism:**
```
[569,121 us] chain [before_dispatch] Elixir.MyApp.Commanded.Support.Middleware.Validate
[23,926  us] chain [after_dispatch] Elixir.Commanded.Middleware.ConsistencyGuarantee
[7,150   us] Aggregate.persist_events.append_to_stream 
```

**100x parallelism:**
```
[6,767,052 us] chain [before_dispatch] Elixir.MyApp.Commanded.Support.Middleware.Validate
[10,437    us] chain [after_dispatch] Elixir.Commanded.Middleware.ConsistencyGuarantee
[3,491     us] Aggregate.persist_events.append_to_stream 
```

It is interesting that `append_to_stream` always stayed around the same value of 3500 uSec, no matter the parallelism - EventStore definitely was not the culprit.

The calls to the `ConsistencyGuarantee` middleware got a little slower, but that may have been due to an increased demand on read model projections (or there may be something else which I'm not sure of).

But the biggest culprit was the `Validate` middleware (which I got from the "Building Conduit" book), and all it does is simply call `Vex.valid?` on the dispatched command. On average it took about 6s (!) per dispatch, and the duration on the `Validate` middleware call was also getting increased (ie 10000us, then 11000us, then 13000us, and so on), which tells me that there's a lock situation going on. I haven't looked into Vex's internals, but it seems to me that there may be a `GenServer` that is a bottleneck, and all of the parallel processes are making synchronous calls to the same `GenServer` (I may be wrong on this one).

The Vex validators I use are just pure functions without any side effects, so I'm not sure what's going on with Vex.

The next obvious step was to get rid of the Vex middleware, and I got the following results by dispatching 1000 commands, with 100 parallel processes:
- With Vex: 75s
- Without Vex: 4s
- Sharding into 10 Commanded Apps/10 Event Stores: 2.4s
- Sharding into 50 Commanded Apps/50 Event Stores: 1.7s

**Summary**
- The biggest benefit came from removing Vex - almost a 20x improvement! I was thinking about replacing it with using Ecto's embedded schemas for doing command validations anyways.
- Running more parallel Commanded Apps/Event Stores helped a little, but not as much as removing Vex.

**Comment by yordis** [comment]
> The biggest benefit came from removing Vex - almost a 20x improvement! I was thinking about replacing it with using Ecto's embedded schemas for doing command validations anyways.

Related to the topic, I intentionally didn't opt-in to use Middleware for validations of any kind.

For us set up.

Most things come from GraphQL, so most value objects are validated at that level. In other cases, we want to validate, we would opt-in for some factory functions:

```elixir
# notice the new!/1 factory function
dispatch(CommandName.new!(%{ ... }))

# or
with {:ok, command} = CommandName.new(%{ ... }), do: dispatch(command)
```

It has its downsides (sometimes somebody forgets to use the correct factory function), but I believe it is a better approach to validating in the middleware. But I think it is worth it in general.

So, here are some take aways from our experience related to validation layers:

1. Validate as much as you can in layers such as GraphQL or the API, forcing the messages to be parsed into reach Value Objects that you can trust
2. Built a suite of factory functions for some value objects that may end up in primitive type anyway
3. Add some factory function for the command at the application-level
4. Don't validate if you feel confident about your data ingestion

**Comment by pcarvsilva** [comment]
@yordis I've did the same as you but i've buld a wrapper over the dispatcher that let's me test if the command was dispatched with the correct struct.

Something like:

`params |> Dispatcher.dispatch(MyCommand)`

 
 Then you can use mox to check if the command was  dispatched with the correct module, also you can use an Ecto Changeset for building the command and checking if it i's valid

**Comment by yordis** [comment]
Related to the validation, here is the existing solution: https://hexdocs.pm/one_piece_commanded/OnePiece.Commanded.Command.html

The `Command` is built on top of https://hexdocs.pm/one_piece_commanded/OnePiece.Commanded.ValueObject.html, which adds the `new/1` and `changeset/2` optional callbacks
